[SYCL][PI][CUDA] Fix context creation property parsing

bjoernknafla opened the pull request:
Align the property parsing of `piContextCreate` with the way OpenCL
properties are provided, i.e., a property list contains keys/IDs and
values, never just a key/ID.

Furthermore enable property lists that just contain a list terminating
`0`.

Introduce a PI type for context properties to untie from OpenCL.

Adapt the SYCL runtime and tests accordingly.

Signed-off-by: Bjoern Knafla <bjoern@codeplay.com>

Commit:
[SYCL] Update libclc install rules

Have libclc install clc-* and libspirv-* to lib and share

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

fwyzard committed:
[SYCL][CUDA] Inline cl namespace to simplify SYCL API usage

Synchronise the CUDA backend with the general SYCL changes from #974.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

Commit:
[SYCL][LIBCLC] Additional libclc builtins to support SYCL work

Adds builtins to libclc to support the CUDA backend for SYCL.

Contributors
Alexander Johnston <alexander@codeplay.com>
David Wood <david.wood@codeplay.com>
Victor Lomuller <victor@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] CMake and lit support for SYCL CUDA backend

Adds defines CMake and lit variables used for SYCL CUDA backend
development and test

Contributors
Alexander Johnston <alexander@codeplay.com>
Bjoern Knafla <bjoern@codeplay.com>
Ruyman Reyes <ruyman@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Local Accessor Support for CUDA

Provides the LocalAccessorToSharedMemory compiler pass required
for supporting SYCL local accessors in CUDA.

Contributors
Alexander Johnston <alexander@codeplay.com>
David Wood <david.wood@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL][CUDA] Change __spirv_BuiltIn.. to functions

Changes the following builtins to functions

__spirv_BuiltInGlobalSize
__spirv_BuiltInWorkgroupSize
__spirv_BuiltInNumWorkgroups
__spirv_BuiltInLocalInvocationId
__spirv_BuiltInWorkgroupId
__spirv_BuiltInGlobalOffset

Contributors
David Wood <david.wood@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
Added missing flags for device-side builtins

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL][CUDA] Removing unnecessary tool from the tree

Acked-by: Victor Lomuller <victor@codeplay.com>
Signed-off-by: Ruyman <ruyman@codeplay.com>

Commit:
[SYCL][CUDA] Fix context creation property parsing

Align the property parsing of `piContextCreate` with the way OpenCL
properties are provided, i.e., a property list contains keys/IDs and
values, never just a key/ID.

Furthermore enable property lists that just contain a list terminating
`0`.

Introduce a PI type for context properties to untie from OpenCL.

Adapt the SYCL runtime and tests accordingly.

Signed-off-by: Bjoern Knafla <bjoern@codeplay.com>

Commit:
[SYCL][PI] Fix kernel group info parameter conversion

Signed-off-by: Steffen Larsen <steffen.larsen@codeplay.com>

fwyzard committed:
[SYCL][CUDA] Refactor __SYCL_INLINE macro

Synchronise the CUDA backend with the general SYCL changes from #1121.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

stuartdadams commented:
LGTM :+1:

Ruyk commented on sycl/plugins/cuda/pi_cuda.cpp:
```diff
@@ -1114,31 +1132,51 @@ pi_result cuda_piContextCreate(const cl_context_properties *properties,
+      // immediately as we want to forge context switches.
+      CUcontext Ctxt;
+      errcode_ret = PI_CHECK_ERROR(cuDevicePrimaryCtxRetain(
+                                     &Ctxt, (*devices)->cuDevice_));
```
```suggestion
                                     &Ctxt, devices[0]->cuDevice_));
```

stuartdadams replied:
Good spot

Ruyk commented on sycl/unittests/pi/cuda/test_base_objects.cpp:
```diff
@@ -118,6 +119,52 @@ TEST_F(DISABLED_CudaBaseObjectsTest, piContextCreatePrimary) {
+  pi_platform platform;
+  pi_device device;
+
+  ASSERT_EQ((Plugins[0].call_nocheck<detail::PiApiKind::piPlatformsGet>(
```
What happens if there is more than one plugin? I guess it doesn't matter at the moment, because the unit tests are disabled, but we need to identify if the plugin is the CUDA one or not, this is OK for later.

stuartdadams replied:
I guess we have a global index somewhere in the tests for controlling which plugin to select

bjoernknafla replied:
This needs fixing but would be more work than this MR, so I adapted to what is already there.

Ruyk replied:
Makes sense, lets just make a note to fix later.

Ruyk reviewed (approved):
Only a minor suggestion, LGTM

Commit:
[SYCL] Have default_selector consider SYCL_BE

Have the default_selector consider the env var SYCL_BE when rating
device scores to make choosing a backend easier.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Select GlobalPlugin based on SYCL_BE

Rather than choose the last found plugin as GlobalPlugin, select
it depending on the SYCL_BE env var.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Improve default device selection checks

Better checks for CUDA and OpenCL devices to match with SYCL_BE in the
default device selection, based on the platform version info.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Formatting update for device_selector.cpp

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL][CUDA] Add SYCL CUDA support to clang driver

Adds CUDA support for sycl compilation in the clang driver

Contributors
Alexander Johnston <alexander@codeplay.com>
David Wood <david.wood@codeplay.com>
Victor Lomuller <victor@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Changed CUDA unit tests to call through plugin

Signed-off-by: Steffen Larsen <steffen.larsen@codeplay.com>

Commit:
[SYCL][CUDA] Initial Implementation of the CUDA backend

Contributors
Alan Forbes <alan.forbes@codeplay.com>
Alexander Johnston <alexander@codeplay.com>
Bjoern Knafla <bjoern@codeplay.com>
Daniel Soutar <daniel.soutar@codeplay.com>
David Wood <david.wood@codeplay.com>
Kumudha Narasimhan <kumudha.narasimhan@codeplay.com>
Mehdi Goli <mehdi.goli@codeplay.com>
Przemek Malon <przemek.malon@codeplay.com>
Ruyman Reyes <ruyman@codeplay.com>
Stuart Adams <stuart.adams@codeplay.com>
Svetlozar Georgiev <svetlozar.georgiev@codeplay.com>
Steffen Larsen <steffen.larsen@codeplay.com>
Victor Lomuller <victor@codeplay.com>

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL][CUDA] Code style and cleanup to CUDA support

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

bjoernknafla commented:
Closing this PR to open a new one rebased on top of upstream `sycl` targeting `intel/llvm`.
